File: 19_pinn/problems/00_pinn/pinn.f90
module pinn_mod
  real(8), parameter :: PI = 3.14159265358979323846d0
contains
  ! 状態を持たない乱数 (パラメータ初期化用): (seed,k) から [0,1)。
  function draw_rand01(seed, k) result(u)
    integer(8), intent(in) :: seed, k
    real(8) :: u
    integer(8), parameter :: M = 2147483647_8
    integer(8) :: x
    x = modulo(modulo(seed, M) * 2654435761_8 + modulo(k, M) + 1_8, M)
    x = modulo(ieor(x, ishft(x, -16)) * 1812433253_8, M)
    x = modulo(ieor(x, ishft(x, -13)) * 1664525_8,    M)
    x = modulo(ieor(x, ishft(x, -16)), M)
    u = real(x, 8) / real(M, 8)
  end function draw_rand01

  ! 損失をパラメータ配列 p (長さ 3H: a=p(1..H), w=p(H+1..2H), b=p(2H+1..3H)) の
  ! 純関数として計算する (スレッドセーフ)。xs は格子点 (長さ M)。
  function loss_fn(H, p, M, xs, lambda_bc) result(L)
    integer, intent(in) :: H, M
    real(8), intent(in) :: p(3*H), xs(M), lambda_bc
    real(8) :: L, res, upp, x, z, t, s2, f, r, u0, u1
    integer :: i, k
    res = 0.0d0
    do i = 1, M
       x = xs(i)
       upp = 0.0d0
       do k = 1, H
          z = p(H+k)*x + p(2*H+k)
          t = tanh(z)
          s2 = 1.0d0 - t*t
          upp = upp + p(k) * p(H+k)*p(H+k) * (-2.0d0*t*s2)
       end do
       f = PI*PI*sin(PI*x)
       r = -upp - f
       res = res + r*r
    end do
    res = res / real(M, 8)
    u0 = 0.0d0; u1 = 0.0d0
    do k = 1, H
       u0 = u0 + p(k) * tanh(p(2*H+k))
       u1 = u1 + p(k) * tanh(p(H+k) + p(2*H+k))
    end do
    L = res + lambda_bc * (u0*u0 + u1*u1)
  end function loss_fn
end module pinn_mod

! 物理情報ニューラルネット (PINN) で -u''(x)=f(x), u(0)=u(1)=0 を「NNで」解く。
! 1隠れ層・tanh なので u'' は解析的 (autodiff 不要)。厳密解 u*=sin(pi x), f=pi^2 sin(pi x)。
! パラメータについての差分勾配の 3H 個の評価を parallel do で分担する。
! 差分法(B系)・CG(G1)と同じ問題を全く別の道具(機械学習)で解く, 最上級の発展課題。
program pinn
  use pinn_mod
  use omp_lib
  character(len=32) :: arg
  integer :: H, M, steps, it, j, q, k, NP, NC, i
  real(8) :: lr, lambda_bc, eps, lp, lm, L, x, u, e, maxerr, t0, elapsed
  real(8), allocatable :: xs(:), p(:), grad(:), pp(:)

  H = 16; M = 50; steps = 4000; lr = 0.01d0
  if (command_argument_count() >= 1) then
     call get_command_argument(1, arg); read (arg, *) H
  end if
  if (command_argument_count() >= 2) then
     call get_command_argument(2, arg); read (arg, *) M
  end if
  if (command_argument_count() >= 3) then
     call get_command_argument(3, arg); read (arg, *) steps
  end if
  if (command_argument_count() >= 4) then
     call get_command_argument(4, arg); read (arg, *) lr
  end if
  lambda_bc = 10.0d0; eps = 1.0d-5; NP = 3*H

  ! コロケーション点 x_i = (i-0.5)/M を (0,1) に等間隔配置
  allocate(xs(M))
  do i = 1, M
     xs(i) = (real(i,8) - 0.5d0) / real(M, 8)
  end do

  ! パラメータ p(3H) = {a, w, b} を小さな乱数で初期化
  allocate(p(NP), grad(NP))
  do k = 1, H
     p(k)     = (draw_rand01(int(k-1,8), 1_8) - 0.5d0) * 0.2d0   ! a
     p(H+k)   = (draw_rand01(int(k-1,8), 2_8) - 0.5d0) * 4.0d0   ! w
     p(2*H+k) = (draw_rand01(int(k-1,8), 3_8) - 0.5d0) * 2.0d0   ! b
  end do

  t0 = omp_get_wtime()
  do it = 0, steps - 1
     ! パラメータについての差分勾配。各 j は独立 (loss は純関数なので各スレッドが
     ! 自分のコピー pp を摂動して評価する)。
     ! TODO: 3H 個のパラメータの差分勾配ループを !$omp parallel do private(j,q,pp,lp,lm) で並列化せよ (各反復で loss_fn を 2 回呼ぶ)。
     do j = 1, NP
        allocate(pp(NP))
        do q = 1, NP
           pp(q) = p(q)
        end do
        pp(j) = p(j) + eps; lp = loss_fn(H, pp, M, xs, lambda_bc)
        pp(j) = p(j) - eps; lm = loss_fn(H, pp, M, xs, lambda_bc)
        grad(j) = (lp - lm) / (2.0d0 * eps)
        deallocate(pp)
     end do
     ! TODO: 上で始めた parallel do 領域を閉じる (!$omp end parallel do)。
     do j = 1, NP
        p(j) = p(j) - lr * grad(j)
     end do

     if (mod(it, 1000) == 0 .or. it == steps - 1) then
        L = loss_fn(H, p, M, xs, lambda_bc)
        print "(a,i4,a,es13.6)", "step ", it, ": loss=", L
     end if
  end do
  elapsed = omp_get_wtime() - t0

  ! 検算: 学習した u(x) を厳密解 sin(pi x) と比べる
  maxerr = 0.0d0; NC = 101
  do i = 0, NC - 1
     x = real(i, 8) / real(NC - 1, 8)
     u = 0.0d0
     do k = 1, H
        u = u + p(k) * tanh(p(H+k)*x + p(2*H+k))
     end do
     e = abs(u - sin(PI*x))
     if (e > maxerr) maxerr = e
  end do
  print "(a,i0,a,i0,a,i0,a,f0.4)", "H=", H, ", M=", M, ", steps=", steps, ", lr=", lr
  print "(a,es12.4)", "final max|u - sin(pi x)| = ", maxerr
  print "(a,f0.3,a)", "elapsed = ", elapsed, " sec"
end program pinn

File: 19_pinn/problems/00_pinn/pinn.cpp
#include <cstdio>
#include <cstdlib>
#include <cmath>
#include <omp.h>

/* 状態を持たない乱数 (パラメータ初期化用): (seed,k) から [0,1)。 */
static inline double draw_rand01(long long seed, long long k) {
  const long long M = 2147483647LL;
  long long x = ((seed % M) * 2654435761LL + (k % M) + 1) % M;
  x = ((x ^ (x >> 16)) * 1812433253LL) % M;
  x = ((x ^ (x >> 13)) * 1664525LL)    % M;
  x =  (x ^ (x >> 16)) % M;
  return (double)x / (double)M;
}

/* 物理情報ニューラルネット (PINN) で微分方程式
     -u''(x) = f(x),  u(0)=u(1)=0
   を「ニューラルネット」で解く。差分法(B系)・CG(G1)と同じ問題を機械学習で解く。

   1隠れ層・tanh のネットなので u の微分は解析的に書ける (autodiff 不要):
     u(x)   = Σ_k a_k tanh(z_k),         z_k = w_k x + b_k
     u'(x)  = Σ_k a_k w_k   sech^2(z_k),  sech^2 = 1 - tanh^2
     u''(x) = Σ_k a_k w_k^2 (-2 tanh(z_k) sech^2(z_k))

   厳密解 u*(x) = sin(pi x) (u(0)=u(1)=0 を満たす) を仕掛けると f(x) = pi^2 sin(pi x)。
   損失 = (1/M) Σ_i ( -u''(x_i) - f(x_i) )^2 + lambda_bc ( u(0)^2 + u(1)^2 )。
   パラメータ {a_k, w_k, b_k} (3H 個) を勾配降下で学習する。
   勾配はパラメータについての差分 (有限差分) で求める:
     grad_j = ( loss(p + eps e_j) - loss(p - eps e_j) ) / (2 eps)。
   この 3H 個の評価は互いに独立なので, パラメータ番号 j で並列化できる。 */

static const double PI = 3.14159265358979323846;

/* 損失をパラメータ配列 p (長さ 3H: a[0..H-1], w[H..2H-1], b[2H..3H-1]) の純関数として
   計算する (スレッドセーフ)。x は格子点 (長さ M)。 */
static double loss_fn(int H, const double * p, int M, const double * xs, double lambda_bc) {
  const double * a = p;
  const double * w = p + H;
  const double * b = p + 2*H;
  /* PDE 残差項 */
  double res = 0.0;
  for (int i = 0; i < M; i++) {
    double x = xs[i];
    double upp = 0.0;        /* u''(x) */
    for (int k = 0; k < H; k++) {
      double z = w[k]*x + b[k];
      double t = tanh(z);
      double s2 = 1.0 - t*t;            /* sech^2 */
      upp += a[k] * w[k]*w[k] * (-2.0*t*s2);
    }
    double f = PI*PI*sin(PI*x);         /* f = -u*'' = pi^2 sin(pi x) */
    double r = -upp - f;
    res += r*r;
  }
  res /= (double)M;
  /* 境界項: u(0), u(1) */
  double u0 = 0.0, u1 = 0.0;
  for (int k = 0; k < H; k++) {
    u0 += a[k] * tanh(b[k]);
    u1 += a[k] * tanh(w[k] + b[k]);
  }
  return res + lambda_bc * (u0*u0 + u1*u1);
}

int main(int argc, char ** argv) {
  int    H     = (argc > 1 ? atoi(argv[1]) : 16);     /* 隠れユニット数 */
  int    M     = (argc > 2 ? atoi(argv[2]) : 50);     /* コロケーション点数 */
  int    steps = (argc > 3 ? atoi(argv[3]) : 4000);   /* 勾配降下ステップ数 */
  double lr    = (argc > 4 ? atof(argv[4]) : 0.01);   /* 学習率 */
  double lambda_bc = 10.0;
  double eps = 1e-5;
  int P = 3*H;                                        /* パラメータ総数 */

  /* コロケーション点 x_i = (i+0.5)/M を (0,1) に等間隔配置 */
  double * xs = (double *)malloc(sizeof(double) * M);
  for (int i = 0; i < M; i++) xs[i] = (i + 0.5) / M;

  /* パラメータ p[3H] = {a, w, b} を小さな乱数で初期化 */
  double * p    = (double *)malloc(sizeof(double) * P);
  double * grad = (double *)malloc(sizeof(double) * P);
  for (int k = 0; k < H; k++) {
    p[k]       = (draw_rand01(k, 1) - 0.5) * 0.2;   /* a */
    p[H+k]     = (draw_rand01(k, 2) - 0.5) * 4.0;   /* w */
    p[2*H+k]   = (draw_rand01(k, 3) - 0.5) * 2.0;   /* b */
  }

  double t0 = omp_get_wtime();
  for (int it = 0; it < steps; it++) {
    /* パラメータについての差分勾配。各 j は独立 (loss は純関数なので各スレッドが
       自分のコピーを摂動して評価する)。 */
    // TODO: 3H 個のパラメータについての差分勾配ループを #pragma omp parallel for で並列化せよ (各反復で loss() を 2 回呼ぶ).
    for (int j = 0; j < P; j++) {
      double * pp = (double *)malloc(sizeof(double) * P);
      for (int q = 0; q < P; q++) pp[q] = p[q];
      pp[j] = p[j] + eps; double lp = loss_fn(H, pp, M, xs, lambda_bc);
      pp[j] = p[j] - eps; double lm = loss_fn(H, pp, M, xs, lambda_bc);
      grad[j] = (lp - lm) / (2.0 * eps);
      free(pp);
    }
    for (int j = 0; j < P; j++) p[j] -= lr * grad[j];

    if (it % 1000 == 0 || it == steps - 1) {
      double L = loss_fn(H, p, M, xs, lambda_bc);
      printf("step %4d: loss=%.6e\n", it, L);
    }
  }
  double elapsed = omp_get_wtime() - t0;

  /* 検算: 学習した u(x) を厳密解 sin(pi x) と比べる */
  const double * a = p; const double * w = p + H; const double * b = p + 2*H;
  double maxerr = 0.0;
  int NC = 101;
  for (int i = 0; i < NC; i++) {
    double x = (double)i / (NC - 1);
    double u = 0.0;
    for (int k = 0; k < H; k++) u += a[k] * tanh(w[k]*x + b[k]);
    double e = fabs(u - sin(PI*x));
    if (e > maxerr) maxerr = e;
  }
  printf("H=%d, M=%d, steps=%d, lr=%g\n", H, M, steps, lr);
  printf("final max|u - sin(pi x)| = %.4e\n", maxerr);
  printf("elapsed = %.3f sec\n", elapsed);
  free(xs); free(p); free(grad);
  return 0;
}
